### ISSUE TITLE
[Feature request] Add support of icingadb-web

### ISSUE THREAD

**Issue by Tontonitch** [HenriWahl/Nagstamon#822]
Icingadb-web module for icingaweb2 has been officially released in version 1.0.0 after a long wait.
I've checked its compatibility with nagstamon, and it seems not compatible.

This icingaweb2 module is intended to replace the "monitoring" module provided with icingaweb2, and based on the old ido db.  
![image](https://user-images.githubusercontent.com/600025/177709696-14cc4732-ec80-4d29-a1c6-6da3d587bf0c.png)

The "monitoring" module should not stay in parallel of the "icingadb/icingadb-web" module, as their have separate configurations with different syntaxes and maintaining both is difficult and won't be possible in the future when icingadb-web will evolve. That's why you can see it as disabled on my screenshot.

The migration page of icingadb-web explain the differences between monitoring and icingadb-web, and in particular the url rewrite that is required to access the same data.
https://icinga.com/docs/icinga-db-web/latest/doc/10-Migration/

Hope that the support of icingadb-web will arrive in nagstamon one time :-)

**Comment by MarcusCaepio** [comment]
@HenriWahl That shouldn't be a huge problem. The URLs and some filter have changed. /monitoring/ is replaced by /icingadb/ and the filters are smarter. I have tried to create a new IcingawebDBServer class, which was just a copy of IcingaWeb2 with adjusted URLs. But I couldn't test it in my environment. Should I just push the class to my fork and you will have a look?

Cheers,
Marcus

**Comment by HenriWahl** [comment]
Icinga already has a lot of modules in Nagstamon. Do you think it might be possible to add a detection into some existing like IcingaWeb2 or are the concepts too different?

**Comment by MarcusCaepio** [comment]
Imho, the concepts are too different. The "old" IDO backend and the "new" IcingaDB backend are totally independent from each other and can also run in parallel. Both backends have their own frontends ( monitoring for IDO, icingdbweb for icingadb). In IcingaDB, the health check is also different, there are two checks now:
![image](https://user-images.githubusercontent.com/7324088/183345670-c1de2b98-f3b8-4cea-9ab1-5076c5d32036.png)

And also all URLs and filter have changed,e.g.:
```
self.cgiurl_services = {'hard': self.monitor_cgi_url + '/icingadb/services?service.state.is_problem=y&service.state.in_downtime=n&service.state.state_type=hard&columns=service.state.last_update,service.state.is_reachable&format=json', \
                                    'soft': self.monitor_cgi_url + '/icingadb/services?service.state.is_problem=y&service.state.in_downtime=n&service.state.state_type=soft&columns=service.state.last_update,service.state.is_reachable&format=json'}

```
In addition, support for the IDO backend will be discontinued in the medium term. So it is easier to handle both in separate modules and later delete the old module, when it's EOL.

**Comment by HenriWahl** [comment]
OK, sounds sound.

**Comment by Tontonitch** [comment]
@MarcusCaepio If it can help I may be able to build and test a windows version on my environment, so if you commit your changes somewhere on your fork I could give it a try and give some feedbacks.

**Comment by HenriWahl** [comment]
@MarcusCaepio just open a pull request.

**Comment by MarcusCaepio** [comment]
sorry all, was on vacation. Will create a PR this week, but it will not contain the new health check, only the icingadb URLs

**Comment by MarcusCaepio** [comment]
@HenriWahl Would you have a look on https://github.com/MarcusCaepio/Nagstamon/blob/IcingaDB/Nagstamon/Servers/IcingaDBWeb.py and tell me, into which of your branches I should create the PR?

**Comment by HenriWahl** [comment]
Just fork from master and add your pull request.

**Comment by Tontonitch** [comment]
Just a quick feedback after the generation of the latest dev build, that should include the https://github.com/HenriWahl/Nagstamon/commit/1aa53d91671b6361a2daa90b6f8dead286b8240d commit.

I was expecting a new "Monitor type", but there is no IcingaDBWeb or similar.
So no way to switch to the added code.

**Comment by HenriWahl** [comment]
Has to be added https://github.com/HenriWahl/Nagstamon/blob/master/Nagstamon/Servers/__init__.py#L35

**Comment by Tontonitch** [comment]
Ok, quickly did that change: https://github.com/HenriWahl/Nagstamon/pull/846

**Comment by Tontonitch** [comment]
Still not able to see the server type in the dropdown list after my change and your fix, but figured out that it was also missing in the code in the server_list variable. Added by https://github.com/HenriWahl/Nagstamon/pull/847

By the way, I've played with the debug mode and it seems to make the app crashing when activated and adding a new server.

**Comment by HenriWahl** [comment]
Please check latest build from master, did some more fix.

**Comment by Tontonitch** [comment]
Ok, I see you've completed the list in the meantime. I will give the latest dev release a try tomorrow ;-)

**Comment by Tontonitch** [comment]
Tested the latest version "Nagstamon-3.9-20220928-win64" and facing the error:

```
DEBUG: 2022-09-29 08:51:38.099306 Monitor server FetchURL: http://monitorsrv2-d/icingaweb2/authentication/login CGI Data: None
DEBUG: 2022-09-29 08:51:38.194386 Monitor server FetchURL: http://monitorsrv2-d/icingaweb2/authentication/login CGI Data: {'redirect': '', 'formUID': 'form_login', 'CSRFToken': '1872105934|1b6a94c9e6eb6427039816064b1d49bdd8766b23de00691d8473f5a01e7ce112', 'btn_submit': 'Login', 'username': 'xxxxx', 'password': 'xxxxx'}
DEBUG: 2022-09-29 08:51:38.377621 Monitor server FetchURL: http://monitorsrv2-d/icingaweb2/icingadb/hosts?host.state.is_problem=y&host.state.state_type=hard&columns=host.state.last_update&format=json CGI Data: None
ERROR: 2022-09-29 08:51:38.407316 Monitor server Traceback (most recent call last):
  File "Nagstamon\Servers\IcingaDBWeb.py", line 203, in _get_status
KeyError: 'host_display_name'
```

**Comment by MarcusCaepio** [comment]
I am going to have a look

**Comment by MarcusCaepio** [comment]
So it looks like not only the URLs have changed, but also the Keys. Sorry, I did not notice this before.
A excerpt of the JSON for 
`/icingadb/hosts?host.state.is_problem=y&host.state.state_type=hard&columns=host.state.last_update&format=json`
contains the following keys with the hostname as value:
```	
"name":"test",
"name_ci":"test",
"display_name":"test",
```
This also belongs to services. A JSON of 
`/icingadb/services/service.state.is_problem=y&service.state.in_downtime=n&service.state.state_type=hard&columns=service.state.last_update,service.state.is_reachable&format=json` 
contents
```
"name":"my service",
"name_ci":"my service",
"display_name":"my service",
```

**Comment by Tontonitch** [comment]
It is a bit more complicate compared to what icingaweb2 returned through the "monitoring" module.B
Below are an example of a host and an example of a service. So it is not a "flat" list.
IcingaDBWeb.py needs more adaptations.
I could put my hands on that, but I suppose @MarcusCaepio you are already working on the required adaptations.

Hosts

```
[
  {
    "id": "54480885747b3a0884b8d9a259e1bf3af57ac3b2",
    "environment_id": "69f7345b830b7397f1ae2623a49b629abd9adfde",
    "name_checksum": "fb55a97ca95584afd595925fbb830f3d423160fe",
    "properties_checksum": "05c5046d81aae3d75e56aa711a73a2a3c61ec324",
    "name": "OTHERNODE",
    "name_ci": "OTHERNODE",
    "display_name": "OTHERNODE",
    "address": "XXX.XXX.XXX.XXX",
    "address6": "",
    "address_bin": null,
    "address6_bin": null,
    "checkcommand_name": "oracle_health",
    "checkcommand_id": "4a534099e79c6567efa59b327237f6cb144dfe89",
    "max_check_attempts": "3",
    "check_timeperiod_name": "24x7",
    "check_timeperiod_id": "3ffaed36caf603d23b5d708635f9ef3f6915d5da",
    "check_timeout": "180",
    "check_interval": "300",
    "check_retry_interval": "120",
    "active_checks_enabled": "1",
    "passive_checks_enabled": "1",
    "event_handler_enabled": "1",
    "notifications_enabled": "1",
    "flapping_enabled": "1",
    "flapping_threshold_low": "25",
    "flapping_threshold_high": "30",
    "perfdata_enabled": "y",
    "eventcommand_name": "",
    "eventcommand_id": null,
    "is_volatile": "",
    "action_url_id": null,
    "notes_url_id": null,
    "notes": "",
    "icon_image_id": "585e46522d7d947e9a96b10fabc44fad66b4f270",
    "icon_image_alt": "",
    "zone_name": "master",
    "zone_id": "2eedbc3883a60948879871c941799fa260c7cadd",
    "command_endpoint_name": "",
    "command_endpoint_id": null,
    "state": {
      "host_id": "54480885747b3a0884b8d9a259e1bf3af57ac3b2",
      "environment_id": "69f7345b830b7397f1ae2623a49b629abd9adfde",
      "state_type": "hard",
      "soft_state": "1",
      "hard_state": "1",
      "previous_soft_state": "99",
      "previous_hard_state": "99",
      "check_attempt": "1",
      "severity": "2112",
      "output": "CRITICAL - cannot connect to OTHERNODE. ORA-12529: TNS:connect request rejected based on current filtering rules (DBD ERROR: OCIServerAttach)",
      "long_output": null,
      "performance_data": null,
      "normalized_performance_data": null,
      "check_commandline": "'/usr/lib/nagios/plugins/check_oracle_health' '--connect' 'OTHERNODE' '--critical' '30' '--mode' 'connection-time' '--password' 'XXXXXXXXXXXXXXX' '--report' 'long' '--username' 'XXXXXXXXXXXXXXX' '--warning' '15'",
      "is_problem": "1",
      "is_handled": "",
      "is_reachable": "1",
      "is_flapping": "",
      "is_overdue": "",
      "is_acknowledged": "",
      "acknowledgement_comment_id": null,
      "last_comment_id": null,
      "in_downtime": "",
      "execution_time": "0.272",
      "latency": "0.001",
      "check_timeout": "180000",
      "check_source": "monitorsrv2-d",
      "scheduling_source": "monitorsrv2-d",
      "last_update": "1664445996.342",
      "last_state_change": "0",
      "next_check": "1664446341.546",
      "next_update": "1664446642.092",
      "last_comment": {
        "environment_id": null,
        "object_type": null,
        "host_id": null,
        "service_id": null,
        "name_checksum": null,
        "properties_checksum": null,
        "name": null,
        "author": null,
        "text": null,
        "entry_type": null,
        "entry_time": null,
        "is_persistent": null,
        "is_sticky": null,
        "expire_time": null,
        "zone_id": null
      }
    },
    "icon_image": {
      "id": "585e46522d7d947e9a96b10fabc44fad66b4f270",
      "icon_image": "custom/logos/vendors2/oracle-db-logo1.png",
      "environment_id": "69f7345b830b7397f1ae2623a49b629abd9adfde"
    }
  },
]

```

Services

```
[
  {
    "id": "0c3758411dff25718a6fa44935698d720c5f349f",
    "environment_id": "69f7345b830b7397f1ae2623a49b629abd9adfde",
    "name_checksum": "5a01dda9cc8468994931d21a6243dad16fe02382",
    "properties_checksum": "8b9aa36f9fad7be8d775046b4d88806389431ae2",
    "host_id": "8af24539731c515a77241576d6d42a32fcf3c399",
    "name": "Alert log",
    "name_ci": "Alert log",
    "display_name": "Alert log",
    "checkcommand_name": "oracle_alertlog",
    "checkcommand_id": "64b11f52652ba6f7ccb2ff8fabd828131570dde5",
    "max_check_attempts": "1",
    "check_timeperiod_name": "24x7",
    "check_timeperiod_id": "3ffaed36caf603d23b5d708635f9ef3f6915d5da",
    "check_timeout": "180",
    "check_interval": "300",
    "check_retry_interval": "120",
    "active_checks_enabled": "1",
    "passive_checks_enabled": "1",
    "event_handler_enabled": "1",
    "notifications_enabled": "1",
    "flapping_enabled": "",
    "flapping_threshold_low": "25",
    "flapping_threshold_high": "30",
    "perfdata_enabled": "y",
    "eventcommand_name": "",
    "eventcommand_id": null,
    "is_volatile": "1",
    "action_url_id": null,
    "notes_url_id": null,
    "notes": "",
    "icon_image_id": null,
    "icon_image_alt": "",
    "zone_name": "master",
    "zone_id": "2eedbc3883a60948879871c941799fa260c7cadd",
    "command_endpoint_name": "",
    "command_endpoint_id": null,
    "state": {
      "service_id": "0c3758411dff25718a6fa44935698d720c5f349f",
      "environment_id": "69f7345b830b7397f1ae2623a49b629abd9adfde",
      "state_type": "hard",
      "soft_state": "3",
      "hard_state": "3",
      "previous_soft_state": "0",
      "previous_hard_state": "0",
      "check_attempt": "1",
      "severity": "1088",
      "output": "Can't locate object method \"addevent\" via package \"Nagios::CheckLogfiles::Search::Oraclealertlog\" at /usr/lib/nagios/plugins/check_logfiles line 5277.",
      "long_output": null,
      "performance_data": null,
      "normalized_performance_data": null,
      "check_commandline": "'/usr/lib/nagios/plugins/check_logfiles' '--config' '/etc/check_logfiles.d/logfiles_oracle_db.cfg' '--environment' 'ORACLE_HOME=/opt/oracle/instantclient_21_6' '--environment
' 'LD_LIBRARY_PATH=/opt/oracle/instantclient_21_6:/usr/local/lib' '--environment' 'TNS_ADMIN=/opt/oracle/instantclient_21_6/network/admin/tnsnames.ora' '--report' 'long' '--searches' 'alertlog' '-t' '18
0'",
      "is_problem": "1",
      "is_handled": "1",
      "is_reachable": "",
      "is_flapping": "",
      "is_overdue": "",
      "is_acknowledged": "",
      "acknowledgement_comment_id": null,
      "last_comment_id": null,
      "in_downtime": "",
      "execution_time": "0.701",
      "latency": "0.001",
      "check_timeout": "180000",
      "check_source": "monitorsrv2-d",
      "scheduling_source": "monitorsrv2-d",
      "last_update": "1664453824.799",
      "last_state_change": "1664453824.799",
      "next_check": "1664454120.519",
      "next_update": "1664454421.926",
      "last_comment": {
        "environment_id": null,
        "object_type": null,
        "host_id": null,
        "service_id": null,
        "name_checksum": null,
        "properties_checksum": null,
        "name": null,
        "author": null,
        "text": null,
        "entry_type": null,
        "entry_time": null,
        "is_persistent": null,
        "is_sticky": null,
        "expire_time": null,
        "zone_id": null
      }
    },
    "host": {
      "id": "8af24539731c515a77241576d6d42a32fcf3c399",
      "environment_id": "69f7345b830b7397f1ae2623a49b629abd9adfde",
      "name_checksum": "3d8ca966853bc114cff27ba550a0588c7aaa38e6",
      "properties_checksum": "cb6b05ce87fb0fbea854279c905de0d313131488",
      "name": "MYDB",
      "name_ci": "MYDB",
      "display_name": "MYDB",
      "address": "XXX.XXX.XXX.XXX",
      "address6": "",
      "address_bin": null,
      "address6_bin": null,
      "checkcommand_name": "oracle_health",
      "checkcommand_id": "4a534099e79c6567efa59b327237f6cb144dfe89",
      "max_check_attempts": "3",
      "check_timeperiod_name": "24x7",
      "check_timeperiod_id": "3ffaed36caf603d23b5d708635f9ef3f6915d5da",
      "check_timeout": "180",
      "check_interval": "300",
      "check_retry_interval": "120",
      "active_checks_enabled": "1",
      "passive_checks_enabled": "1",
      "event_handler_enabled": "1",
      "notifications_enabled": "1",
      "flapping_enabled": "1",
      "flapping_threshold_low": "25",
      "flapping_threshold_high": "30",
      "perfdata_enabled": "y",
      "eventcommand_name": "",
      "eventcommand_id": null,
      "is_volatile": "",
      "action_url_id": null,
      "notes_url_id": null,
      "notes": "",
      "icon_image_id": "585e46522d7d947e9a96b10fabc44fad66b4f270",
      "icon_image_alt": "",
      "zone_name": "master",
      "zone_id": "2eedbc3883a60948879871c941799fa260c7cadd",
      "command_endpoint_name": "",
      "command_endpoint_id": null,
      "state": {
        "host_id": "8af24539731c515a77241576d6d42a32fcf3c399",
        "environment_id": "69f7345b830b7397f1ae2623a49b629abd9adfde",
        "state_type": "hard",
        "soft_state": "1",
        "hard_state": "1",
        "previous_soft_state": "0",
        "previous_hard_state": "0",
        "check_attempt": "1",
        "severity": "2112",
        "output": "CRITICAL - cannot connect to MYDB. ORA-12537: TNS:connection closed (DBD ERROR: OCIServerAttach)",
        "long_output": null,
        "performance_data": null,
        "normalized_performance_data": null,
        "check_commandline": "'/usr/lib/nagios/plugins/check_oracle_health' '--connect' 'MYDB' '--critical' '30' '--mode' 'connection-time' '--password' 'XXXXXXXXXXXXXXX' '--report' 'long' '--username' 'XXXXXXXXXXXXXXX' '--warning' '15'",
        "is_problem": "1",
        "is_handled": "",
        "is_reachable": "1",
        "is_flapping": "",
        "is_overdue": "",
        "is_acknowledged": "",
        "acknowledgement_comment_id": null,
        "last_comment_id": null,
        "in_downtime": "",
        "execution_time": "0.304",
        "latency": "0.001",
        "check_timeout": "180000",
        "check_source": "monitorsrv2-d",
        "scheduling_source": "monitorsrv2-d",
        "last_update": "1664453866.154",
        "last_state_change": "1664453629.674",
        "next_check": "1664454152.095",
        "next_update": "1664454452.706"
      }
    },
    "icon_image": {
      "id": null,
      "icon_image": null,
      "environment_id": null
    }
  },
]

```

**Comment by MarcusCaepio** [comment]
@Tontonitch you are welcome to take over. When I created my PR, I hoped it's enough to change the URLs, because that was only necessary in the WebUI. I did not know, they changed the whole JSON.

**Comment by Tontonitch** [comment]
Ok, I will work on my fork, test and PR when ready.

**Comment by Tontonitch** [comment]
I've a version that works, but it needs more work before a PR.
@MarcusCaepio you can retrieve from https://github.com/Tontonitch/Nagstamon/releases/tag/latest and test it if you want.
From my quick checks, remaining fixes to do:
* "SERVICE UNREACHABLE" printed at the end of service output
* redirections url to icingaweb2 not correct
* acknowledgements not working

**Comment by Tontonitch** [comment]
@MarcusCaepio you can open issues there if you test the latest app release: https://github.com/Tontonitch/Nagstamon/issues
@HenriWahl I don't think the app is supposed to crash in case the acknowledgement is not effective. And it prevent me to have usefull info in the app logs. I've pasted the eventviewer entry of the crash there: https://github.com/Tontonitch/Nagstamon/issues/2

**Comment by HenriWahl** [comment]
@Tontonitch could you run it from source in a terminal window so there will be some error message why it crashes?

**Comment by Tontonitch** [comment]
Thanks for the tip, I thought it would be more difficult to launch it from sources on Windows, but that was easy.
I've progressed, the form on icingadb is different and no form name nor id, I had to rely on class name.
I'm able to send the ack request with the expected form data, but it is not taken into account by icinga after. I need to investigate more.

**Comment by Tontonitch** [comment]
I've fixed the acknowledgment part. Same logic will be applied to the other parts (recheck,...) so that's progressing well.

**Comment by Tontonitch** [comment]
Ok, I've finished a good release candidate. Tested everything, works fine. PR: https://github.com/HenriWahl/Nagstamon/pull/853.

**Comment by HenriWahl** [comment]
Included in latest build 3.9-20221010.
